validator/server_arb/redis/consumer.go

```diff
@@ -0,0 +1,115 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/pubsub"
+	"github.com/offchainlabs/nitro/util/redisutil"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	"github.com/offchainlabs/nitro/validator"
+	"github.com/offchainlabs/nitro/validator/server_api"
+	"github.com/spf13/pflag"
+)
+
+type ExecutionSpawnerConfig struct {
+	RedisURL       string                `koanf:"redis-url"`
+	ConsumerConfig pubsub.ConsumerConfig `koanf:"consumer-config"`
+	// Supported wasm module roots.
+	ModuleRoots []string `koanf:"module-roots"`
+}
+
+var DefaultExecutionSpawnerConfig = ExecutionSpawnerConfig{
+	RedisURL:       "",
+	ConsumerConfig: pubsub.DefaultConsumerConfig,
+	ModuleRoots:    []string{},
+}
+
+var TestExecutionSpawnerConfig = ExecutionSpawnerConfig{
+	RedisURL:       "",
+	ConsumerConfig: pubsub.TestConsumerConfig,
+	ModuleRoots:    []string{},
+}
+
+func ExecutionSpawnerConfigAddOptions(prefix string, f *pflag.FlagSet) {
+	pubsub.ConsumerConfigAddOptions(prefix+".consumer-config", f)
+	f.StringSlice(prefix+".module-roots", nil, "Supported module root hashes")
+}
+
+func (cfg *ExecutionSpawnerConfig) Enabled() bool {
+	return cfg.RedisURL != ""
+}
+
+type ExecutionSpawner struct {
+	stopwaiter.StopWaiter
+	spawner validator.ExecutionSpawner
+
+	// consumers stores moduleRoot to consumer mapping.
+	consumers map[common.Hash]*pubsub.Consumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash]
+}
+
+func NewExecutionSpawner(cfg *ExecutionSpawnerConfig, spawner validator.ExecutionSpawner) (*ExecutionSpawner, error) {
+	if cfg.RedisURL == "" {
+		return nil, fmt.Errorf("redis url cannot be empty")
+	}
+	redisClient, err := redisutil.RedisClientFromURL(cfg.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	consumers := make(map[common.Hash]*pubsub.Consumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash])
+	for _, hash := range cfg.ModuleRoots {
+		mr := common.HexToHash(hash)
+		c, err := pubsub.NewConsumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash](redisClient, server_api.RedisStreamForRoot(mr), &cfg.ConsumerConfig)
```

Comment:
Have a separate stream for BOLD-execution (don't use the same as validation)
OTOH - Can probably re-use a lot of the logic from validation to connect to stream, timeout, etc..

Reply:
Done.